lib/AddonSignals.py

```diff
@@ -1,135 +1,161 @@
+# GNU Lesser General Public License v2.1 (see COPYING or https://www.gnu.org/licenses/gpl-2.1.txt)
+"""The AddonSignals module provides signal/slot mechanism for inter-addon communication in Kodi"""
 
-import sys
-import base64
-import json
-import xbmc
-import xbmcaddon
+from xbmc import executeJSONRPC, log, LOGNOTICE, Monitor, sleep
 
-RECEIVER = None
 
+def _addon_id():
+    """Return the Kodi add-on id and cache it as a static variable"""
+    if not hasattr(_addon_id, 'cached'):
+        from xbmcaddon import Addon
+        _addon_id.cached = Addon().getAddonInfo('id')
+    return getattr(_addon_id, 'cached')
 
-def _getReceiver():
-    global RECEIVER  # pylint: disable=global-statement
-    if not RECEIVER:
-        RECEIVER = SignalReceiver()
-    return RECEIVER
 
-
-def _decodeData(data):
+def _decode_data(data):
+    """Decode base64-encoded JSON data and return Python data structure"""
+    import json
     encoded_data = json.loads(data)
-    if encoded_data:
-        json_data = base64.b64decode(encoded_data[0])
-        # NOTE: With Python 3.5 and older json.loads() does not support bytes or bytearray
-        if isinstance(json_data, bytes):
-            json_data = json_data.decode('utf-8')
-        return json.loads(json_data)
-
-    return None
-
-
-def _encodeData(data):
+    if not encoded_data:
+        return None
+    from base64 import b64decode
+    json_data = b64decode(encoded_data[0])
+    # NOTE: With Python 3.5 and older json.loads() does not support bytes or bytearray
+    return json.loads(_to_unicode(json_data))
+
+
+def _encode_data(data):
+    """Encode Python data structure into base64-encoded JSON data"""
+    from base64 import b64encode
+    import json
     json_data = json.dumps(data)
     if not isinstance(json_data, bytes):
         json_data = json_data.encode('utf-8')
-    encoded_data = base64.b64encode(json_data)
-    if sys.version_info[0] > 2:
-        encoded_data = encoded_data.decode('ascii')
-    return encoded_data
+    encoded_data = b64encode(json_data)
+    return encoded_data.decode('ascii')
+
+
+def _receiver():
+    """Return a SignalReceiver instance and cache it as a static variable"""
+    if not hasattr(_receiver, 'cached'):
+        _receiver.cached = SignalReceiver()
+    return getattr(_receiver, 'cached')
 
 
 def _jsonrpc(**kwargs):
-    ''' Perform JSONRPC calls '''
+    """Perform JSONRPC calls"""
+    import json
     if 'id' not in kwargs:
-        kwargs.update(id=1)
+        kwargs.update(id=0)
     if 'jsonrpc' not in kwargs:
         kwargs.update(jsonrpc='2.0')
-    return json.loads(xbmc.executeJSONRPC(json.dumps(kwargs)))
+    return json.loads(executeJSONRPC(json.dumps(kwargs)))
 
 
-class SignalReceiver(xbmc.Monitor):
+def _to_unicode(text, encoding='utf-8', errors='strict'):
+    """Force text to unicode"""
+    if isinstance(text, bytes):
+        return text.decode(encoding, errors=errors)
+    return text
+
+
+class SignalReceiver(Monitor, object):
```

Comment:
why add object?

Reply:
Because we need new-style classes in Python 2 in order to use `__subclasses__`. For Python 3 it wouldn't be needed as it would be a new-style class anyway.

Reply:
This is only needed for integration testing though, unfortunately.